src/memory_prefetch.c

```diff
@@ -0,0 +1,410 @@
+    uint64_t key_hash;        /* Hash value of the key being prefetched */
+    dictEntry *current_entry; /* Pointer to the current entry being processed */
+} PrefetchInfo;
+
+/* CommandsBatch structure holds the state of the current batch of client commands being processed. */
+typedef struct CommandsBatch {
+    size_t cur_idx;              /* Index of the current key being processed */
+    size_t keys_done;            /* Number of keys that have been prefetched */
+    size_t key_count;            /* Number of keys in the current batch */
+    size_t client_count;         /* Number of clients in the current batch */
+    size_t max_prefetch_size;    /* Maximum number of keys to prefetch in a batch */
+    size_t executed_commands;    /* Number of commands executed in the current batch */
+    int *slots;                  /* Array of slots for each key */
+    void **keys;                 /* Array of keys to prefetch in the current batch */
+    client **clients;            /* Array of clients in the current batch */
+    dict **keys_dicts;           /* Main dict for each key */
+    dict **expire_dicts;         /* Expire dict for each key */
+    dict **current_dicts;        /* Points to either keys_dicts or expire_dicts */
+    PrefetchInfo *prefetch_info; /* Prefetch info for each key */
+} CommandsBatch;
+
+static CommandsBatch *batch = NULL;
+
+void freePrefetchCommandsBatch(void) {
+    if (batch == NULL) {
+        return;
+    }
+
+    zfree(batch->clients);
+    zfree(batch->keys);
+    zfree(batch->keys_dicts);
+    zfree(batch->expire_dicts);
+    zfree(batch->slots);
+    zfree(batch->prefetch_info);
+    zfree(batch);
+    batch = NULL;
+}
+
+void prefetchCommandsBatchInit(void) {
+    serverAssert(!batch);
+    size_t max_prefetch_size = server.prefetch_batch_max_size;
+
+    if (max_prefetch_size == 0) {
+        return;
+    }
+
+    batch = zcalloc(sizeof(CommandsBatch));
+    batch->max_prefetch_size = max_prefetch_size;
+    batch->clients = zcalloc(max_prefetch_size * sizeof(client *));
+    batch->keys = zcalloc(max_prefetch_size * sizeof(void *));
+    batch->keys_dicts = zcalloc(max_prefetch_size * sizeof(dict *));
+    batch->expire_dicts = zcalloc(max_prefetch_size * sizeof(dict *));
+    batch->slots = zcalloc(max_prefetch_size * sizeof(int));
+    batch->prefetch_info = zcalloc(max_prefetch_size * sizeof(PrefetchInfo));
+}
+
+void onMaxBatchSizeChange(void) {
+    if (batch && batch->client_count > 0) {
+        /* We need to process the current batch before updating the size */
+        return;
+    }
+
+    freePrefetchCommandsBatch();
+    prefetchCommandsBatchInit();
+}
+
+/* Prefetch the given pointer and move to the next key in the batch. */
+static void prefetch(void *addr) {
+    valkey_prefetch(addr);
+    /* While the prefetch is in progress, we can continue to the next key */
+    batch->cur_idx = (batch->cur_idx + 1) % batch->key_count;
+}
+
+static void markDone(PrefetchInfo *info) {
```

Comment:
```suggestion
static void markKeyAsFetched(PrefetchInfo *info) {
```
or `markKeyAsdone`.

Reply:
Changed to `markKeyAsDone` as we didn't necessarily prefetch it.